design/repository-maintenance.md

```diff
@@ -0,0 +1,236 @@
+
+3. **Update Maintenance Job Result in BackupRepository CR**: Retrieve the result of the maintenance job and update the status of the `BackupRepository` CR accordingly.
+
+4. **Add Setting for Maintenance Job**: Introduce a configuration option to set maintenance jobs, including resource limits (CPU and memory), keeping the latest N failed maintenance jobs.
+
+4. **Clear Up Maintenance Jobs**: Introduce a configuration to determine the number of failed maintenance jobs to retain for debugging purposes.
+
+
+## Detailed Design
+
+### 1. Add Maintenance CLI Command
+
+The CLI command will be added to the Velero CLI.
+
+```shell
+$ velero maintenance --repo-name $repo-name --repo-type $repo-type --backup-storage-location $bsl
+```
+
+The main maintenance logic would be using the repository provider to do the maintenance.
+```golang
+func (o *Options) Run(f veleroCli.Factory) {
+	...
+	credentialFileStore, err := credentials.NewNamespacedFileStore(
+		cli,
+		f.Namespace(),
+		"/tmp/credentials",
+		filesystem.NewFileSystem(),
+	)
+	cmd.CheckError(err)
+	credentialSecretStore, err := credentials.NewNamespacedSecretStore(cli, f.Namespace())
+	cmd.CheckError(err)
+	// Initialize repo provider
+	repoProvider := provider.NewUnifiedRepoProvider(
+		credentials.CredentialGetter{
+			FromFile:   credentialFileStore,
+			FromSecret: credentialSecretStore,
+		}, o.RepoType, cli, logrus.New())
+	
+	// Get backupRepository
+	repo, err := repository.GetBackupRepository(context.Background(), cli, f.Namespace(),
+		repository.BackupRepositoryKey{
+			VolumeNamespace: o.VolumeNamespace,
+			BackupLocation:  o.BackupStorageLocation,
+			RepositoryType:  o.RepoType,
+		}, true)
+	cmd.CheckError(err)
+	// Get BSL
+	bsl := &velerov1api.BackupStorageLocation{}
+	cli.Get(context.Background(), client.ObjectKey{Namespace: f.Namespace(), Name: repo.Spec.BackupStorageLocation}, bsl)
+	cmd.CheckError(err)
+
+	para := provider.RepoParam{
+		BackupRepo:     repo,
+		BackupLocation: bsl,
+	}
+	// Connect
+	err = repoProvider.BoostRepoConnect(context.Background(), para)
+	cmd.CheckError(err)
+	// Prune repo
+	err = repoProvider.PruneRepo(context.Background(), para)
+	cmd.CheckError(err)
+	...
+}
+```
+
+### 2. Create Jobs by Repository Manager
+Currently, the backup repository controller will call the repository manager to do the `PruneRepo`, and Kopia or Restic maintenance is then finally called through multiple chain calls.
+
+We will keep using the `PruneRepo` function in the repository manager, but we cut off the multiple chain calls by creating a maintenance job.
+
+```golang
+func (m *manager) PruneRepo(repo *velerov1api.BackupRepository) error {
+	...
+	// writing the maintenance result into default terminationMessagePath
+	args := []string {
+	"-c",
+	"velero maintenance --repo-name=" + param.BackupRepo.Spec.VolumeNamespace + " --repo-type=" + param.BackupRepo.Spec.RepositoryType + " --backup-storage-location=" + param.BackupLocation.Name + " > /dev/termination-log",
```

Comment:
We should only put the final error message to file `/dev/termination-log` and put the msg in the repo status then.
The whole logs of the maintenance job can be got via the job logs.

Reply:
yes, you are right